promql/engine.go

```diff
@@ -2141,3 +2302,128 @@ func unwrapParenExpr(e *parser.Expr) {
 		}
 	}
 }
+
+// WrapWithStepInvariantExpr wraps all possible parts of the given
+// expression with StepInvariantExpr wherever valid.
+func WrapWithStepInvariantExpr(expr parser.Expr) parser.Expr {
+	isStepInvariant := wrapWithStepInvariantExprHelper(expr)
+	if isStepInvariant {
+		return newStepInvariantExpr(expr)
+	}
+	return expr
+}
+
+// wrapWithStepInvariantExprHelper wraps the child nodes of the expression
+// with a StepInvariantExpr wherever valid. The returned boolean is true if the
+// passed expression qualifies to be wrapped by StepInvariantExpr.
+func wrapWithStepInvariantExprHelper(expr parser.Expr) bool {
+	switch n := expr.(type) {
+	case *parser.VectorSelector:
+		if n.Timestamp > 0 {
+			return true
+		}
+
+	case *parser.AggregateExpr:
+		// TODO(codesome): check Param?
+		e := wrapWithStepInvariantExprHelper(n.Expr)
+		if e {
+			return true
+		}
+		return false
+
+	case *parser.BinaryExpr:
+		e1, e2 := wrapWithStepInvariantExprHelper(n.LHS), wrapWithStepInvariantExprHelper(n.RHS)
+		if e1 && e2 {
+			return true
+		}
+
+		if e1 {
+			n.LHS = newStepInvariantExpr(n.LHS)
+		}
+		if e2 {
+			n.RHS = newStepInvariantExpr(n.RHS)
+		}
+
+		return false
+
+	case *parser.Call:
```

Comment:
This is wrong, just because a functions arguments are the same over time doesn't mean it's result is the same over time.

Reply:
Can you give example when it won't be same over time?

Reply:
(I see I had to exclude the `time()` function out of it, what else?)

Reply:
Any of the time functions.

In general functions are blackboxes to the engine, you can't make any presumptions about them.

Reply:
> In general functions are blackboxes to the engine, you can't make any presumptions about them.

But since we keep the PromQL execution deterministic, I think it is not wrong to assume that results won't change if input is the same. I can exclude all time functions from here.

Reply:
> But since we keep the PromQL execution deterministic, I think it is not wrong to assume that results won't change if input is the same.

That's not correct. Just because one set of inputs always produces one deterministic output doesn't mean you can vary a given input and get that same output.

> I can exclude all time functions from here.

No, as I said functions are blackboxes. Their names should not appear in this file.

Keep in mind also that there's other users of this code who will inject their own functions which we know nothing about. Not that we officially support them, but having code which it is impossible for them to deal with correctly doesn't seem wise.

Reply:
Coming back to this function call discussion

> Their names should not appear in this file.

I would add a function to `parser.Call` struct called `IsStepInvariant` which will would tell if it remains same for same data at different step (hence all time functions returning false). I guess that should be enough, Talking about names appearing in this file, I already see a special handling in place for `"timestamp"` which contradicts what you said. 

> Keep in mind also that there's other users of this code who will inject their own functions which we know nothing about.

We don't need to know too. If a user is trying to inject something into the PromQL, they are solely responsible for it since you already said we don't officially support them. I think adding `IsStepInvariant` to the `Call` struct should simplify it anyway.

Reply:
> I would add a function to parser.Call struct called IsStepInvariant

The parser doesn't know about function semantics, this would be a matter for the function map.

> Talking about names appearing in this file, I already see a special handling in place for "timestamp" which contradicts what you said.

Indeed, but that proceeded the function list being public and is exactly one unusual function - not an entire class of functions. I suspect even with this PR you're going to need more special casing of timestamp.